pkg/minikube/machine/cache_images.go

```diff
@@ -326,6 +330,173 @@ func removeExistingImage(r cruntime.Manager, src string, imgName string) error {
+
+	succeeded := []string{}
+	failed := []string{}
+
+	for _, p := range profiles { // loading images to all running profiles
+		pName := p.Name // capture the loop variable
+
+		c, err := config.Load(pName)
+		if err != nil {
+			// Non-fatal because it may race with profile deletion
+			klog.Errorf("Failed to load profile %q: %v", pName, err)
+			failed = append(failed, pName)
+			continue
+		}
+
+		for _, n := range c.Nodes {
+			m := config.MachineName(*c, n)
+
+			status, err := Status(api, m)
+			if err != nil {
+				klog.Warningf("error getting status for %s: %v", m, err)
+				failed = append(failed, m)
+				continue
+			}
+
+			if status == state.Running.String() { // the not running hosts will load on next start
+				h, err := api.Load(m)
+				if err != nil {
+					klog.Warningf("Failed to load machine %q: %v", m, err)
+					failed = append(failed, m)
+					continue
+				}
+				cr, err := CommandRunner(h)
+				if err != nil {
+					return err
+				}
+				if cacheDir != "" {
+					// saving image names, to cache
+					err = SaveCachedImages(c, cr, images, cacheDir)
+				} else {
+					// saving mage files
+					err = SaveLocalImages(c, cr, images, output)
+				}
+				if err != nil {
+					failed = append(failed, m)
+					klog.Warningf("Failed to load cached images for profile %s. make sure the profile is running. %v", pName, err)
+					continue
+				}
+				succeeded = append(succeeded, m)
+			}
+		}
+	}
+
+	klog.Infof("succeeded pulling from : %s", strings.Join(succeeded, " "))
+	klog.Infof("failed pulling from : %s", strings.Join(failed, " "))
+	// Live pushes are not considered a failure
+	return nil
+}
+
+// transferAndSaveCachedImage transfers and loads a single image from the cache
+func transferAndSaveCachedImage(cr command.Runner, k8s config.KubernetesConfig, imgName string, cacheDir string) error {
+	dst := filepath.Join(cacheDir, imgName)
+	dst = localpath.SanitizeCacheDir(dst)
+	return transferAndSaveImage(cr, k8s, dst, imgName)
+}
+
+// transferAndSaveImage transfers and loads a single image
+func transferAndSaveImage(cr command.Runner, k8s config.KubernetesConfig, dst string, imgName string) error {
+	r, err := cruntime.New(cruntime.Config{Type: k8s.ContainerRuntime, Runner: cr})
+	if err != nil {
+		return errors.Wrap(err, "runtime")
+	}
+
+	klog.Infof("Saving image to: %s", dst)
+	filename := filepath.Base(dst)
+
+	_, err = os.OpenFile(dst, os.O_CREATE|os.O_WRONLY, 0777)
+	if err != nil {
+		return err
+	}
+
+	f, err := assets.NewFileAsset(dst, saveRoot, filename, "0644")
+	if err != nil {
+		return errors.Wrapf(err, "creating copyable file asset: %s", filename)
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			klog.Warningf("error closing the file %s: %v", f.GetSourcePath(), err)
+		}
+	}()
+
+	src := path.Join(saveRoot, filename)
+	args := append([]string{"rm", "-f"}, src)
+	if _, err := cr.RunCmd(exec.Command("sudo", args...)); err != nil {
+		return err
+	}
+	err = r.SaveImage(imgName, src)
```

Comment:
@afbjorklund i think that for saving image to `saveRoot` == `/var/lib/minikube/images` to work, cruntime.SaveImage has to be called as root - eg, for docker:
https://github.com/kubernetes/minikube/blob/c3c4d0455dfed89650fdf54f9f70d551912b4969/pkg/minikube/cruntime/docker.go#L227

Reply:
Yes, I think this should be done in the SaveImages. Maybe use a temp file, maybe use a different directory.

It was hidden by the fact that we run the other clients (ctr) or programs (podman) as root, unlike docker client.

Reply:
Used pipes instead, since I didn't want to call the docker client with `sudo`.